cmd/enter: honor user’s configured shell inside toolbox

yann-soubeyrand opened the pull request.

HarryMichal commented:
@yann-soubeyrand, this is marvellous! This is something I've wanted to do for Toolbx in the near future. It'll still have to wait a little bit before merging (and most definitely will mess the nearest release) but I'm quite certain this is exactly what we need. Many thanks!

spencergilbert commented on doc/toolbox-sh.1.md:
```diff
@@ -0,0 +1,21 @@
+% toolbox-sh(1)
+
+## NAME
+toolbox\-init\-container - Initialize a running container
```
```suggestion
toolbox\-sh - Launch user configured shell inside container
```

yann-soubeyrand commented:
Hello @HarryMichal, do you have any news about this PR?

debarshiray commented on src/cmd/sh.go:
```diff
@@ -0,0 +1,96 @@
+
+	s := strings.Split(strings.TrimSuffix(string(out), "\n"), ":")[6]
+
+	err = syscall.Exec(s, []string{"-l"}, os.Environ())
```
We will eventually need a fallback mechanism here.

Even today, if the user's shell on the *host* is Z shell, then right after `toolbox create`, the container is unlikely to have `/bin/zsh` in it.  So, `toolbox enter` will show an error message, but still fall back to `/bin/bash`, to let the user enter the container and install Z shell there.  From there on, `toolbox enter` will keep using `/bin/zsh`.

It will be good to eventually fold that into this new command, so that the fallback mechanism doesn't need to go through several invocations of `podman exec`, which is slow.

Since, we are now also looking at the user's default shell *inside* the container, there's a possibility that the shell is not available inside the container.  I agree that this is unlikely, but it's possible that the shell got uninstalled because of a mistake.  If we don't offer a fallback in this case, then it will be quite cumbersome for the user to recover the container.  The default shell on the host could be a fallback option.

Note that `toolbox enter` is meant to be used interactively, so it's different from something like [nologin(8)](https://man7.org/linux/man-pages/man8/nologin.8.html).

Anyway, doing all that will require moving around a good amount of code, so it's better to keep it simple in the first commit, and build upon it subsequently.  However, to not break `git bisect`, we need to ensure that even the initial simple implementation uses an exit code of 127 to keep the existing (slow) fallback working.

yann-soubeyrand replied:
I’m thinking about the way to implement the fallback mechanism, it’s not ready at the time.

yann-soubeyrand replied:
I think I’ve a working fallback mechanism.

debarshiray commented on src/cmd/initContainer.go:
```diff
@@ -405,7 +393,6 @@ func configureUsers(targetUserUid int,
 			"--append",
 			"--groups", sudoGroup,
 			"--home", targetUserHome,
-			"--shell", targetUserShell,
```
We still need to set up the user's default shell inside the container to be something sane as the initial starting point or *factory state*.  Otherwise, for the vast majority of users, who don't change their default shell and just use Bash, the default shell inside the container will become `/bin/sh`.  That will be disturbing.

The most sane default inside the container is to match the host, because very few people will want separate shells across their containers.

We need to figure out a way to persistently record (across reboots) that the container was already set up as per this factory state, and that everything from now on is a local change made by the user. This way, `toolbox init-container` can avoid overwriting the user's local changes.

Removing this persistent record (whatever it is) should reset the container back to its factory state. This can be very helpful to debug problems and/or recover broken containers by going back to a known-good state.

yann-soubeyrand replied:
I restored the `--shell` option from `toolbox init-container` as you requested.

debarshiray reviewed (changes requested):
Thanks for working on this, @yann-soubeyrand !

debarshiray commented on doc/toolbox-sh.1.md:
```diff
@@ -0,0 +1,21 @@
+% toolbox-sh(1)
+
+## NAME
+toolbox\-sh - Launch user configured shell inside container
```
There are other use-cases for this: https://github.com/containers/toolbox/issues/988

It can have an impact on the name.  Although, since it's an internal implementation detail we can always change the name of the command later.

yann-soubeyrand commented on src/cmd/sh.go:
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright © 2019 – 2021 Red Hat Inc.
```
Fixed, I anticipated a bit on 2023 since there’s only a few days left :wink:

yann-soubeyrand commented on doc/toolbox-create.1.md:
```diff
@@ -138,4 +138,6 @@ $ toolbox create --authfile ~/auth.json --image registry.example.com/bar
 ## SEE ALSO
 
-`toolbox(1)`, `toolbox-init-container(1)`, `podman(1)`, `podman-create(1)`, `podman-login(1)`, `podman-pull(1)`, `containers-auth.json(5)`
+`toolbox(1)`, `toolbox-init-container(1)`, `toolbox-sh(1)`,
```
Fixed.

yann-soubeyrand commented on doc/toolbox-sh.1.md:
```diff
@@ -0,0 +1,21 @@
+% toolbox-sh(1)
```
Ditto, fixed.

yann-soubeyrand commented on src/cmd/sh.go:
```diff
@@ -0,0 +1,96 @@
+		return fmt.Errorf("failed to get current user: %w", err)
+	}
+
+	out, err := exec.Command("getent", "passwd", u.Username).Output()
```
Done, I replaced my exec with the `LookupUser` function in the package you linked me to which offers the desired functionality.

yann-soubeyrand commented on src/cmd/sh.go:
```diff
@@ -0,0 +1,96 @@
+		return err
+	}
+
+	s := strings.Split(strings.TrimSuffix(string(out), "\n"), ":")[6]
```
Done (see above).

yann-soubeyrand commented:
Hi @debarshiray,

Regarding initial shell configuration, I was wondering why we don’t let `useradd` use the default system configuration in `/etc/default/useradd`. Indeed, that seems to be the more sensible thing to do since we’d be assured the configured shell would be present (unless the container image used is really bad) and the user shouldn’t be surprised to end up with the default shell configured in the container image.

However, I didn’t understand why my first shell was always configured to `/bin/sh` whereas `useradd` configuration indicates `/bin/bash` as the default. It turns out that my user isn’t created by `toolbox init-container` but by `podman` due to the use of the `--userns keep-id` option.

I’m not sure where to go from now on. Should we try to get `podman` use the default shell configured in the container image? Or should we go with the mechanism to indicate that “factory reset” of the toolbox has already been done?

yann-soubeyrand commented:
> I’m not sure where to go from now on. Should we try to get `podman` use the default shell configured in the container image? Or should we go with the mechanism to indicate that “factory reset” of the toolbox has already been done?

`podman` seems to directly write `passwd` entries in `/etc/passwd` so I guess it won’t be possible to easily and reliably make it use the default shell of the system. We’re left with the second solution then. How do you see it? Should we put a file in `/var/lib/toolbox/` inside the toolbox container to indicate that initialization has already been done?

yann-soubeyrand commented on doc/toolbox-init-container.1.md:
```diff
@@ -102,11 +101,6 @@ The bind mounted paths are:
 - `/var/log/journal`
 - `/var/mnt`
 
-**--shell** SHELL
```
I restored the `--shell` option from `toolbox init-container` to serve as factory reset configuration as you suggested.

yann-soubeyrand commented on src/cmd/initContainer.go:
```diff
@@ -107,15 +106,6 @@ func init() {
 		false,
 		"Ensure that certain configuration files inside the toolbox container are in sync with the host")
 
-	flags.StringVar(&initContainerFlags.shell,
```
See above.

yann-soubeyrand commented:
Hi @debarshiray, I think I’ve addressed your points and I’d be glad to hear your feedback.

I’ve tried to separated the work done by `toolbox init-container` in two categories: the toolbox initialization phase, which should be done only once during the first container start, and the start phase, which should be done each time the container is started. I’m not sure I separated things correctly, though.

I’ve still unanswered questions in my head:
- Will we want to use the `toolbox sh` command for `toolbox run`?
- How does the whole thing with `capsh` work? I’m under the impression that `capsh` execs `bash`, which execs `toolbox sh`, which execs the user shell. Is it possible to simplify things? What’s the role of `capsh` here?
